### PR TITLE
[GPU] rnnt accuracy degradation

### PR DESCRIPTION
**Pull request by kelvinchoi-intel**
### Details:
 - *Gather layer in rnnt prediction model has out of order index and this cause random value for the input*

### Tickets:
 - *106886*

### CODE REVIEW

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/gather_ref.cl [hunk 1]
```diff
@@ -56,6 +56,9 @@ KERNEL(gather_ref)(
 
     INPUT0_TYPE val = dictionary[dictionary_idx];
 
+    if (dictionary_idx >= DICTIONARY_SIZE)
+        val = 0;
+
```
Comment (e-ddykim) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/gather_ref.cl:61]: What do you think of using a short form as below?

```suggestion
    INPUT0_TYPE val = (dictionary_idx < DICTIONARY_SIZE) ? dictionary[dictionary_idx] : 0;
```

Comment (kelvinchoi-intel) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/gather_ref.cl]: fixed

**Review by yeonbok — APPROVED** [review]

**Review by e-ddykim — APPROVED** [review]
Looks good to me

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/gather_ref.cl [hunk 2]
```diff
@@ -54,7 +54,11 @@ KERNEL(gather_ref)(
     const uint dictionary_idx = GET_DICTIONARY_INDEX(DICTIONARY_INDEX_ORDER);
     const uint output_idx = GET_INDEX(OUTPUT,,ORDER);
 
+#if AXIS_DIM
+    INPUT0_TYPE val = (INPUT_AXIS_INDEX >= 0 && INPUT_AXIS_INDEX < AXIS_DIM) ? dictionary[dictionary_idx] : 0;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/gather_ref.cl:58]: shouldn't we support this case in shape agnostic kernel too? 
i.e., shouldnt we check the actual size in shape_info and compare?

**Review by ahnyoung-paul — APPROVED** [review]

File: src/plugins/intel_gpu/src/kernel_selector/kernels/gather/gather_kernel_ref.cpp [hunk 3]
```diff
@@ -211,11 +236,15 @@ CommonDispatchData GatherKernelRef::SetDefault(const gather_params& params) cons
 JitConstants GatherKernelRef::GetJitConstants(const gather_params& params) const {
     JitConstants jit = MakeBaseParamsJitConstants(params);
 
+    jit.AddConstant(MakeJitConstant("GATHER_AXIS_SHAPE_INFO_INDEX", GetGatherAxisIndexInShapeInfo(params)));
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/gather/gather_kernel_ref.cpp:239]: I think we need to add this only when needed, i.e., params.is_shape_agnostic

Comment (kelvinchoi-intel) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/gather/gather_kernel_ref.cpp:239]: done

**Pull request merged**